core/server/common/src/main/java/alluxio/master/journal/ufs/UfsJournalCheckpointThread.java

```diff
@@ -168,12 +174,57 @@ public long getNextSequenceNumber() {
 
   @Override
   public void run() {
+    // Start a new thread which tracks the journal replay statistics
+    ExponentialBackoffRetry retry = new ExponentialBackoffRetry(1000, 30000, Integer.MAX_VALUE);
+    long start = System.currentTimeMillis();
+    final OptionalLong finalSN = UfsJournalReader.getLastSN(mJournal);
+    Thread t = new Thread(() -> {
+      long lastMeasuredSeq = 0;
+      long lastMeasuredTime = start;
+      while (!Thread.currentThread().isInterrupted() && retry.attempt()) {
+        // log current stats
+        long currTime = System.currentTimeMillis();
+        long currSeqNum = mLastAppliedSN; // single volatile read
+        long interval = currTime - lastMeasuredTime;
+        long appliedEntries = currSeqNum - lastMeasuredSeq;
+        double entryRateS = ((double) appliedEntries) / interval;
+
+        StringJoiner logMsg = new StringJoiner("|");
+        logMsg.add(String.format("journal replay on %s", mJournal));
+        logMsg.add(String.format("current SN: %d", currSeqNum));
+        logMsg.add(String.format("entries in last %dms=%d", interval, appliedEntries));
+        if (finalSN.isPresent()) {
+          long remaining = finalSN.getAsLong() - currSeqNum;
+          double remainingTime = ((double) remaining) / entryRateS;
+          if (remaining > 0) {
+            logMsg.add(String.format("est. entries left: %d", remaining));
+          }
+          if (!Double.isNaN(remainingTime) && !Double.isInfinite(remainingTime)) {
+            logMsg.add(String.format("est. time remaining %.2fms", remainingTime));
+          }
+        }
+
+        LOG.info(logMsg.toString());
```

Comment:
Would it be possible to extract the calculation and logging of the replay stats to a util method and reuse it in both UFS and embedded journals?

Reply:
Hmm..I can try this. It might be messy though :upside_down_face:

Reply:
added